fcomm/src/lib.rs

```diff
@@ -687,8 +687,8 @@ impl Opening<Scalar> {
 
             // public_output = (result_expr (secret . new_fun))
             let cons = public_output.expr;
-            let result_expr = s.car(&cons);
-            let new_comm = s.cdr(&cons);
+            let result_expr = s.car(&cons).map_err(|_| Error::EvaluationFailure)?;
+            let new_comm = s.cdr(&cons).map_err(|_| Error::EvaluationFailure)?;
```

Comment:
This seems to show that `EvaluationFailure` is at least misnamed (if it even should be returned). That's because no evaluation is happening here, so it makes no sense that an error indicating failure of (the non-existent) evaluation occurred. The `Store.cdr()` is just looking up the cdr of a `Cons` in a `Store`.

Reply:
Okay, looking more closely, I realize that I misread what's happening here. Sorry. That having been said, I think this is still probably the wrong error type.

Reply:
If we get to this point, then we know for certain that the evaluation that happened above (line 683) *succeeded*.

For some reason (completely concealed by this `map_err`), we're not able to extract the car/cdr of the result. One possible reason is that the function being applied doesn't return a correctly structured output (i.e. `public_output.expr` is not a cons). Or something else might have gone wrong.

Reply:
We're still inside the call to `Evaluator::eval`, right? But I think what you're saying here is we can be more specific. I'm totally on board with that, I just didn't know what exactly was happening here. But now that I do I can try to come up with something better, like `ReturnValueError` or some such.

Reply:
I don't think so… the evaluation should have completed when the call on line 683 completes.

The lines I'm commenting are the ones with the calls `s.car(…)` and `s.cdr(…)`. Those are just calls interrogating the store (`s`), to traverse some value that is hopefully a cons.

Reply:
Ah I see, you are right. I can add a new variant to `fcomm::Error` and use it here, maybe something like `InvalidResult(lurk::error::LurkError)`?

Reply:
Hmmm. It's not really a Lurk error, though — but since we're just trying to wrap an error from `lurk-rs` maybe that works.

Following the structure of the fcomm errors, this is probably just an `OpeningFailure`. In this case, the actual problem is related to the destructuring of the output failing because we are specifically trying to open a chained functional commitment. So the most informative error would highlight this fact.

Based on that, maybe we can add a message to `OpeningFailure` then in this case, create a message which notes the immediate cause (attempted destructuring of chained commitment output) then provides further message context from the underlying error.